docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,251 @@
+* `Stake(Meta, Stake)`: `Delegation` (validator, stake amount, activation / deactivation epochs, warmup rate) + `credits_observed` (used for rewards calc)
+* `RewardsPool`: deprecated
+
+To properly perform a redelegation, we need to store at least the
+next validator's vote pubkey and a redelegation epoch, which means that
+redelegation absolutely requires a larger struct.
+
+### Structure
+
+We need to introduce a new type and an upgrade path from legacy / V1 stakes:
+
+```
+struct StakeV2 {
+    meta: Meta,
+    stake: Stake,
+    last_claim_epoch: Epoch, // needed to know if it's up-to-date, see `ClaimRewards`
+    redelegation: Stake,
+}
+```
+
+With this information, we can calculate multi-epoch redelegations when the cluster
+sees more than 25% stake movement.
+
+### Interoperability with V1
+
+Since loads of places use `mem::size_of::<StakeState>()`, and we still need
+to support that, that struct won't get touched. On the flip side, programs
+need to process both. Since there's already a neat interface on `StakeState`,
+we can do something similar with a new type:
+
+```
+enum StakeStateV2 {
+    Uninitialized,
+    Initialized(Meta),
+    StakeV1(Meta, Stake),
+    RewardsPool,
+    StakeV2(StakeV2),
+}
+impl StakeStateV2 {
+    /// ... all the same as StakeState ...
+    pub fn redelegation(&self) -> Option<Stake> {
+      match self {
+          Self::StakeV2(stake) => Some(stake.redelegation),
+          _ => None,
+      }
+    }
+}
+```
+
+Programs update to use `StakeStateV2` everywhere, and existing code still works.
+When an `Initialized` stake is delegated, the stake program makes it a `StakeV1`
+if the size is `200`, and a `StakeV2` if the size is `256`.
+
+### Epoch Boundary Calculations
+
+While we're at it, new stake accounts can also solve point number 2, by having
+their rewards paid through an update instruction, and not automatically at every
+epoch boundary.
+
+During epoch boundary, the validator vote accounts receive all stake V2
+account rewards as withheld lamports in their account. They cannot
+touch these, and stake accounts may claim them.
+
+The biggest users of stake accounts, ie stake pools and stake bots, already
+touch stake accounts every epoch during their bookkeeping, so this adds trivial
+work for them.
+
+Legacy stake accounts are unaffected, and inflation continues to pay out their rewards
+at epoch boundaries automatically.
+
+## New Stake Program Instructions
+
+`Split` / `Merge` / `Deactivate` are only allowed if the stake account has
+been updated this epoch, or after deactivation epoch if applicable.
+
+Side note: deactivate might actually be OK, since we could technically calculate
+the rewards even after deactivation.
+
+### UpgradeToV2
+
+Given a V1 stake account, a signature from the stake or withdraw authority, and
+a funding account, reallocate the existing stake account and convert it to a V2
+stake. The funding account adds the lamports necessary to cover the new rent-exemption.
```

Comment:
I have a few questions about how migration would work both for native staking and stake pools.

**Native Staking:**
1. Would converting a stake account from V1 to V2 result in any lost epoch rewards?
2. Would it be possible to convert a V2 account back to a V1 account or would the conversion only work one way?

**Stake Pools:**
1. Would the SPL stake-pool program have an instruction added to migrate stake accounts to V2 using the V2 upgrade command?
2. Would the CLI interface for `spl-stake-pool` offer options on whether to use V1 or V2 stake accounts?
3. Would the SPL stake-pool program be kept upgradeable until V2 stake accounts are released to ensure that the program can be upgraded to more seamlessly support migrations from V1 to V2?

Reply:
> 1. Would converting a stake account from V1 to V2 result in any lost epoch rewards?

No, it just rewrites the account

> 2. Would it be possible to convert a V2 account back to a V1 account or would the conversion only work one way?

The general idea is one-way conversion.  I can't think of a good reason to downgrade.  If you really wanted to do that, you'd have to create a new v1 stake account again.

> 1. Would the SPL stake-pool program have an instruction added to migrate stake accounts to V2 using the V2 upgrade command?

Good question. We're probably going to make a v2 stake pool program at some point, so it might be that the v2 stake pool program only uses v2 stake accounts.  If the migration happens earlier, then we'll probably want a way to upgrade.  We'll have to be very careful about the rollout either way.

> 2. Would the CLI interface for spl-stake-pool offer options on whether to use V1 or V2 stake accounts?

If the option exists in the program, yes

> 3. Would the SPL stake-pool program be kept upgradeable until V2 stake accounts are released to ensure that the program can be upgraded to more seamlessly support migrations from V1 to V2?

We're going to freeze the v1 stake pool program soon, once the minimum delegation is released.  That means that v2 stake accounts would only be supported in the v2 stake pool program.  So yes, we'd probably keep it upgradeable until v2 stake accounts are released.